### ISSUE TITLE
Implement 

### ISSUE THREAD

**Issue by neatudarius** [beman-project/Optional26#13]

**Comment by camio** [comment]
@ldionne , you mentioned at C++Now the idea of using the LLVM unit test framework in Beman. Is this something you would still recommend? Does it support this use case? Where can we learn more about it?

**Comment by ldionne** [comment]
@camio Yes, Lit has support for something called `clang-verify`, which allows checking that code fails with a specific diagnostic. In general, note that you don't want to just check that code fails to compile, since that's too brittle (e.g. a typo would render the test useless since it would indeed fail, but not for the reason you expect). You really want to say "this fails for THAT reason" instead.

Clang-verify supports various directives like `expected-error`, `expected-warning`, etc. The full set of directives supported and how to use them is documented in [the Clang documentation](https://clang.llvm.org/docs/InternalsManual.html#specifying-diagnostics).

Here's an example we have in libc++ where we check that a specific instantiation triggers a static assertion with a message (from [here](https://github.com/llvm/llvm-project/blob/56f0ecd6db9219b7d14a8eda613d6b75060643eb/libcxx/test/std/thread/thread.semaphore/ctor.verify.cpp#L17)):

```
#include <semaphore>

void f() {
  std::counting_semaphore<-1> s(2); // expected-error-re@*:* {{static assertion failed{{.*}}The least maximum value must be a positive number}}
  (void)s;
}
```

You can also specify the file / line where the diagnostic should happen, how many times it should happen, etc. It's pretty flexible -- it originated as a way to test Clang itself and we're merely making use of the same functionality in libc++.

**Comment by jwakely** [comment]
I consider it essential to be able to test for specific diagnostics for std::lib code. A static assert that clearly says the template argument must be copy constructible is an important QoI property.

**Comment by tothambrus11** [comment]
That's a very interesting type of test, I think these are pretty much underrated. They would be very useful to test APIs generally, to make sure they cannot be used incorrectly.

My instant reaction was, can we hack concepts to verify these things, as they are able to see if something would compile or not. You could enable each specific test case by compile-time checks based on whether you are compiling on which C++ version, and just try compiling everywhere. But as @ldionne mentioned, this would be too brittle, we should check for the reason - so some preprocessing-based technology like Clang's testing framework would need to be used.

Adapting Clang-verify would be already a huge leap forward but it's a bit suboptimal in terms of usability:
1. We cannot mix normal tests with these compile-time death tests.
2. We cannot instrument different compilation targets easily per test case. AFAIU only one `RUN` command can be set up per file, and then it's globally applied to all targets.
3. There is no autocomplete (and probably no error-checking) on these comment-based test instrumentations. So if we make a mistake in writing a typo in the test case, we might just have an always-passing test. Maybe a macro or `constexpr if` based approach could help.

So if someone has time, it would be great to improve on these areas, and possibly upstream these features to some popular testing framework. It is insightful to look at the design decisions made by [DocTest](https://github.com/doctest/doctest), the fastest testing framework built for TDD.

**Comment by jwakely** [comment]
Concepts can't be used, because there are lots of errors that do not happen in the immediate context of template substitution, so cannot be detected using concepts (or SFINAE equivalents).

For example, `optional::and_then` has:

_Mandates_: `remove_cvref_t<U>` is a specialization of `optional`.

Failing to meet that condition makes a program ill-formed, but you cannot detect it with concepts, because it's enforced with a `static_assert` in the function body **not** with a requires-clause on the function declaration. It cannot be a constraint on the function declaration, because that would cause unwanted instantiations of generic lambdas during overload resolution, which can make valid code ill-formed.

So the reason for not using concepts is not because they don't allow checking for a specific diagnostic message, but because they simply don't work for the general case. Not all errors are template substitution errors.

**Comment by jwakely** [comment]
> - We cannot mix normal tests with these compile-time death tests.  

Why not? Isn't it just a different kind of annotation within the test, so that some tests are expected to compile+link+run without errors, and some tests are expected to fail with a specific error or warning.

> - We cannot instrument different compilation targets easily per test case. AFAIU only one RUN command can be set up per file, and then it's globally applied to all targets.

Would you actually want that for test that check for required errors?

Anyway, DejaGnu supports it, any `{ dg-error ... }` or `{ dg-warning ... }` directive can be target-specific, so you can match different errors for different targets. e.g. there might be a nice user-friendly `static_assert` message for C++11 and later, but just a nasty template instantiation error for C++98 mode.

> - There is no autocomplete (and probably no error-checking) on these comment-based test instrumentations. So if we make a mistake in writing a typo in the test case, we might just have an always-passing test. Maybe a macro or constexpr if based approach could help.

DejaGnu deals with that too. If you write `{ dg-eror ... }` instead of dg-error then the test will fail because it doesn't know what `dg-eror` means. Some typos are not caught, e.g. `dgerror` would be ignored by the test framework, because it only matches `dg-*`. But that's not a huge problem in practice. GCC has tens of thousands of tests spanning several decades and this just isn't a problem we encounter often. And it's pretty easy to create a separate linter (even just based on `grep`) that can check all tests for such problems.

If you had a framework with all the features you want except that typos in specially-formatted comments are not caught automatically, I would not reject that framework.
